main/midi/midiPlayer.c

```diff
@@ -0,0 +1,1194 @@
+static void handleMidiEvent(midiPlayer_t* player, const midiStatusEvent_t* event);
+static void handleSysexEvent(midiPlayer_t* player, const midiSysexEvent_t* sysex);
+static void handleMetaEvent(midiPlayer_t* player, const midiMetaEvent_t* event);
+static void handleEvent(midiPlayer_t* player, const midiEvent_t* event);
+
+static const midiTimbre_t defaultDrumkitTimbre = {
+    .type = NOISE,
+    .flags = TF_PERCUSSION,
+    .percussion = {
+        .playFunc = defaultDrumkitFunc,
+        // TODO: Define the data and put it here!
+        .data = NULL,
+    },
+    .envelope = { 0 },
+    .name = "Swadge Drums 0",
+};
+
+static const midiTimbre_t acousticGrandPianoTimbre = {
+    .type = WAVETABLE,
+    .flags = TF_NONE,
+    .waveIndex = 0,
+    .envelope = {
+        // TODO: Just realized I forgot how ADSR actually works halfway through writing everything else...
+        // So go make sure the rest of everything makes sense, maybe rename everything to {a,d,r}Time and {s}Level for clarity
+        // Pretty fast attack
+        .attack = MS_TO_TICKS(24),
+        // Take a good long while to reach the sustain level
+        .decay = MS_TO_TICKS(750),
+        // Sustain at about 75% of initial volume
+        .sustain = 192,
+        // And a not-too-short release time
+        .release = MS_TO_TICKS(100),
+        // Yup, I'm sure it will sound exactly like a grand piano now!
+    },
+    .name = "Acoustic Grand Piano",
+};
+
+// Check for the first unused note, then try to steal one in order of less to more bad, and return INT32_MAX if none are
+// available
+/**
+ * @brief Return the index of an unallocated voice from the given voice pool.
+ *
+ * This function finds the voice index to allocate, but the caller is responsible for updating the state
+ * bitmaps to actuallymark it as allocated.
+ * If there are no unallocated voices remaining, an allocated voice index may be returned.
+ *
+ * @param states A pointer to the voice state bitmaps for this pool
+ * @param voiceCount The number of voices in this pool
+ * @return uint32_t The index of the voice to allocate
+ */
+static uint32_t allocVoice(const voiceStates_t* states, uint8_t voiceCount)
+{
+    uint32_t allStates
+        = VS_ANY(states)
+          | states->held; // states->attack | states->decay | states->release | states->sustain | states->held;
+
+    // Set up a bitflag which has a 1 set for every voice that is NOT being used
+    //                         /- flip the bits so a 1 represents an unused voice and a 0 represents an in-use voice
+    //                        /                /- mask the bits to only go up to the number of voices we have
+    //                       v                v
+    uint32_t unusedVoices = (~allStates) & (0xFFFFFFFFu >> (32 - voiceCount));
+
+    if (unusedVoices != 0)
+    {
+        // Return whatever the first voice that's not allocated is
+        return __builtin_ctz(unusedVoices);
```

Comment:
Don't see this every day. Not that we were gonna migrate away from `gcc` or anything...

Reply:
Ah yeah, though fwiw it sounds like clang at least also supports this and a bunch other funky gcc builtins

Reply:
`clang` is smart enough to be compatible. `tcc` isn't, but I'm not a `tcc` diehard.